File: egottol/engines/eii/actuators.py
import logging
from typing import Dict, List, Tuple

import numpy as np

from egottol.engines.eii.types import EIIState, ImpulseEvent
from egottol.models.eii import ActuatorMode, FeedbackActuatorConfig

logger = logging.getLogger(__name__)


class FeedbackActuator:
    """Feedback operator Gamma: inference outputs actuate the circuit."""

    def __init__(self, config: FeedbackActuatorConfig):
        self.config = config
        self.mode = config.mode
        self._last_pre_spike_time: Dict[Tuple[int, int], float] = {}

    def actuate(
        self,
        state: EIIState,
        prediction: np.ndarray,
        confidence: float,
        events: List[ImpulseEvent],
        target_class: int = 0,
    ) -> np.ndarray:
        y = np.asarray(prediction, dtype=float).reshape(-1)
        if self.mode == ActuatorMode.DAC:
            control = self._actuate_dac(y)
        elif self.mode == ActuatorMode.STDP:
            control = self._actuate_stdp(state, y, confidence, events, target_class)
        elif self.mode == ActuatorMode.DIGITAL:
            control = self._actuate_digital(state, y)
        elif self.mode == ActuatorMode.OPTICAL:
            control = self._actuate_optical(state, y)
        else:
            logger.warning("Unknown actuator mode %s; falling back to DAC", self.mode)
            control = self._actuate_dac(y)

        state.control = control
        return control

    def _actuate_dac(self, prediction: np.ndarray) -> np.ndarray:
        return self.config.v_dd * np.clip(prediction, 0.0, 1.0)

    def _actuate_stdp(
        self,
        state: EIIState,
        prediction: np.ndarray,
        confidence: float,
        events: List[ImpulseEvent],
        target_class: int,
    ) -> np.ndarray:
        if state.conductances is None:
            return self._actuate_dac(prediction)

        p_correct = float(prediction[target_class % len(prediction)])
        e_inference = 1.0 - p_correct
        control = self._actuate_dac(prediction)

        for event in events:
            key = (event.channel, event.channel)
            dt_spike = state.t - self._last_pre_spike_time.get(key, state.t)
            self._last_pre_spike_time[key] = state.t
            delta_t = dt_spike if event.event_type != "memristor_switch" else 0.0
            delta_g = self._stdp_delta(delta_t) * e_inference * self.config.stdp_eta
            idx = event.channel % state.conductances.size
            state.conductances.flat[idx] = np.clip(
                state.conductances.flat[idx] + delta_g,
                1e-9,
                1.0,
            )
        return control

    def _stdp_delta(self, delta_t: float) -> float:
        if delta_t > 0:
            return self.config.stdp_a_plus * np.exp(-delta_t / max(self.config.stdp_tau_plus, 1e-9))
        if delta_t < 0:
            return -self.config.stdp_a_minus * np.exp(delta_t / max(self.config.stdp_tau_minus, 1e-9))
        return 0.0

    def _actuate_digital(self, state: EIIState, prediction: np.ndarray) -> np.ndarray:
        control = np.zeros_like(prediction)
        for idx, value in enumerate(prediction):
            reg_name = f"OUT_{idx}"
            bit = 1.0 if value >= self.config.digital_threshold else 0.0
            state.logic_registers[reg_name] = bit
            control[idx] = bit
        return control

    def _actuate_optical(self, state: EIIState, prediction: np.ndarray) -> np.ndarray:
        if state.optical_phases is None:
            state.optical_phases = np.zeros(len(prediction), dtype=float)
        delta_phi = self.config.optical_phase_scale * prediction
        state.optical_phases = state.optical_phases + delta_phi
        return delta_phi

    def source_modulation(self, prediction: np.ndarray, t: float) -> float:
        cls = int(np.argmax(prediction))
        amplitude = self.config.source_amplitude_scale * float(prediction[cls])
        omega = 2.0 * np.pi * self.config.source_frequency
        return amplitude * np.sin(omega * t)

File: core/eii/actuators.cpp
#include "actuators.h"

#include <algorithm>
#include <cmath>

namespace deepiri {

std::vector<double> FeedbackActuator::actuate(EIIState& state, const std::vector<double>& prediction,
                                               double confidence, const std::vector<ImpulseEvent>& events,
                                               int targetClass) {
    std::vector<double> control;
    switch (config_.mode) {
        case ActuatorMode::Stdp:
            control = actuateStdp(state, prediction, confidence, events, targetClass);
            break;
        case ActuatorMode::Digital:
            control = actuateDigital(state, prediction);
            break;
        case ActuatorMode::Optical:
            control = actuateOptical(state, prediction);
            break;
        case ActuatorMode::Dac:
        default:
            control = actuateDac(prediction);
            break;
    }
    state.control = control;
    return control;
}

std::vector<double> FeedbackActuator::actuateDac(const std::vector<double>& prediction) const {
    std::vector<double> out(prediction.size());
    for (size_t i = 0; i < prediction.size(); ++i) out[i] = config_.vDd * std::clamp(prediction[i], 0.0, 1.0);
    return out;
}

double FeedbackActuator::stdpDelta(double deltaT) const {
    if (deltaT > 0.0) return config_.stdpAPlus * std::exp(-deltaT / std::max(config_.stdpTauPlus, 1e-9));
    if (deltaT < 0.0) return -config_.stdpAMinus * std::exp(deltaT / std::max(config_.stdpTauMinus, 1e-9));
    return 0.0;
}

std::vector<double> FeedbackActuator::actuateStdp(EIIState& state, const std::vector<double>& prediction,
                                                   double confidence, const std::vector<ImpulseEvent>& events,
                                                   int targetClass) {
    (void)confidence;
    if (state.conductances.empty()) {
        return actuateDac(prediction);
    }
    size_t idxTarget = prediction.empty() ? 0 : static_cast<size_t>(((targetClass % static_cast<int>(prediction.size())) + static_cast<int>(prediction.size())) % static_cast<int>(prediction.size()));
    double pCorrect = prediction.empty() ? 0.0 : prediction[idxTarget];
    double eInference = 1.0 - pCorrect;
    std::vector<double> control = actuateDac(prediction);

    for (const auto& event : events) {
        int key = event.channel;
        double lastT = lastPreSpikeTime_.count(key) ? lastPreSpikeTime_[key] : state.t;
        double dtSpike = state.t - lastT;
        lastPreSpikeTime_[key] = state.t;
        double deltaT = (event.eventType != "memristor_switch") ? dtSpike : 0.0;
        double deltaG = stdpDelta(deltaT) * eInference * config_.stdpEta;
        size_t idx = static_cast<size_t>(((event.channel % static_cast<int>(state.conductances.size())) +
                                           static_cast<int>(state.conductances.size())) %
                                          static_cast<int>(state.conductances.size()));
        state.conductances[idx] = std::clamp(state.conductances[idx] + deltaG, 1e-9, 1.0);
    }
    return control;
}

std::vector<double> FeedbackActuator::actuateDigital(EIIState& state, const std::vector<double>& prediction) const {
    std::vector<double> control(prediction.size(), 0.0);
    for (size_t idx = 0; idx < prediction.size(); ++idx) {
        double bit = prediction[idx] >= config_.digitalThreshold ? 1.0 : 0.0;
        state.logicRegisters["OUT_" + std::to_string(idx)] = bit;
        control[idx] = bit;
    }
    return control;
}

std::vector<double> FeedbackActuator::actuateOptical(EIIState& state, const std::vector<double>& prediction) const {
    if (state.opticalPhases.size() != prediction.size()) {
        state.opticalPhases.assign(prediction.size(), 0.0);
    }
    std::vector<double> deltaPhi(prediction.size());
    for (size_t i = 0; i < prediction.size(); ++i) {
        deltaPhi[i] = config_.opticalPhaseScale * prediction[i];
        state.opticalPhases[i] += deltaPhi[i];
    }
    return deltaPhi;
}

double FeedbackActuator::sourceModulation(const std::vector<double>& prediction, double t) const {
    if (prediction.empty()) return 0.0;
    size_t cls = static_cast<size_t>(std::max_element(prediction.begin(), prediction.end()) - prediction.begin());
    double amplitude = config_.sourceAmplitudeScale * prediction[cls];
    double omega = 2.0 * M_PI * config_.sourceFrequency;
    return amplitude * std::sin(omega * t);
}

}  // namespace deepiri
